### PR TITLE
fluentd logger: support all options besides Unix sockets

### PR DESCRIPTION
**Pull request by pcarrier**
Mostly useful for docker/docker#19438.

### CODE REVIEW

File: daemon/logger/fluentd/fluentd.go [hunk 1]
```diff
@@ -8,11 +8,13 @@ import (
 	"net"
 	"strconv"
 	"strings"
+	"time"
 
 	"github.com/Sirupsen/logrus"
 	"github.com/docker/docker/daemon/logger"
 	"github.com/docker/docker/daemon/logger/loggerutils"
 	"github.com/fluent/fluent-logger-golang/fluent"
+	"github.com/fsouza/go-dockerclient/external/github.com/docker/docker/pkg/units"
```
Comment (wenchma) [inline daemon/logger/fluentd/fluentd.go]: this import is strange, you made a lot of detours.

Comment (pcarrier) [inline daemon/logger/fluentd/fluentd.go]: Sorry, my IDE did something weird. Fixed.

**Comment by calavera** [comment]
Small change, otherwise LGTM

**Comment by pcarrier** [comment]
@calavera addressed, thanks.

**Comment by icecrime** [comment]
Ping @LK4D4!

**Comment by calavera** [comment]
This needs a rebase :sob:

**Comment by pcarrier** [comment]
AFAICT this is because of 3cf82ff1ab14e1ddd2b629524e894ac359168388, which tries to solve the same problem in what I'd say isn't as good a solution. My solution would be to revert it.

File: vendor/src/github.com/fluent/fluent-logger-golang/fluent/fluent.go [hunk 2]
```diff
@@ -194,7 +210,14 @@ func (f *Fluent) close() (err error) {
 
 // connect establishes a new connection using the specified transport.
 func (f *Fluent) connect() (err error) {
-	f.conn, err = net.DialTimeout("tcp", f.Config.FluentHost+":"+strconv.Itoa(f.Config.FluentPort), f.Config.Timeout)
+	switch f.Config.FluentNetwork {
+	case "tcp":
+		f.conn, err = net.DialTimeout(f.Config.FluentNetwork, f.Config.FluentHost+":"+strconv.Itoa(f.Config.FluentPort), f.Config.Timeout)
```
Comment (cpuguy83) [inline vendor/src/github.com/fluent/fluent-logger-golang/fluent/fluent.go]: Can we move this host/port vs sock path into a helper function?
This way connect is just a single case and can potentially support other formats easily as well (e.g. "unixgram")

Comment (cpuguy83) [inline vendor/src/github.com/fluent/fluent-logger-golang/fluent/fluent.go]: For that matter, seems like we should just store host/port as a single field anyway.

Comment (pcarrier) [inline vendor/src/github.com/fluent/fluent-logger-golang/fluent/fluent.go]: This is a third-party library.

**Comment by pcarrier** [comment]
@calavera @icecrime @LK4D4 what next?

File: vendor/src/github.com/fluent/fluent-logger-golang/fluent/fluent.go (hunk 2 above)
Comment (cpuguy83) [inline vendor/src/github.com/fluent/fluent-logger-golang/fluent/fluent.go]: oh damn, didn't notice we were in vendor here.

File: daemon/logger/fluentd/fluentd.go [hunk 3]
```diff
@@ -24,12 +26,24 @@ type fluentd struct {
...
-	defaultHostName                  = "localhost"
-	defaultPort                      = 24224
-	defaultTagPrefix                 = "docker"
-	defaultIgnoreConnectErrorOnStart = false           // So that we do not break existing behaviour
-	defaultBufferLimit               = 1 * 1024 * 1024 // 1M buffer by default
+	name = "fluentd"
+
+	defaultHost        = "127.0.0.1"
+	defaultPort        = 24224
+	defaultBufferLimit = 8 * 1024 * 1024
```
Comment (cpuguy83) [inline daemon/logger/fluentd/fluentd.go]: I'd personally prefer to keep this @ a 1MB buffer.
Consider you have 100 containers (which is not much) each with an 8MB buffer

Comment (pcarrier) [inline daemon/logger/fluentd/fluentd.go]: This would be changing the current buffer size. Happy to adjust, if you confirm we don't want stable "defaults".

Comment (pcarrier) [inline daemon/logger/fluentd/fluentd.go]: Also note that the dehaviour when exceeding the buffer size is pretty damn aggressive: the whole buffer is discarded. Again, third-party library.

**Comment by LK4D4** [comment]
ping @tagomoris

File: daemon/logger/fluentd/fluentd.go (hunk 3 above)
Comment (tagomoris) [inline daemon/logger/fluentd/fluentd.go]: It looks better to keep it as 1MB. Users can change it if they want.

Comment (pcarrier) [inline daemon/logger/fluentd/fluentd.go]: To be clear, the last docker release is at 8MB. I'm OK going with 1MB if we want, but that would be a change.

Comment (tagomoris) [inline daemon/logger/fluentd/fluentd.go]: That is already discussed on #18041.
Please respect previous commits.

File: daemon/logger/fluentd/fluentd.go [hunk 4]
```diff
@@ -114,8 +162,6 @@ func ValidateLogOpt(cfg map[string]string) error {
 		case "tag":
 		case "labels":
 		case "env":
-		case "fail-on-startup-error":
-		case "buffer-limit":
```
Comment (tagomoris) [inline daemon/logger/fluentd/fluentd.go]: Is it ok to just remove these options, without adding options newly added? (`fluentd-buffer-limit`, `fluentd-retry-limit` and others).

File: docs/admin/logging/overview.md [hunk 5]
```diff
@@ -171,17 +171,20 @@ underscore (`_`).
...
 
 `docker run --log-driver=fluentd --log-opt fluentd-address=localhost:24224 --log-opt tag=docker.{{.Name}}`
 
-If container cannot connect to the Fluentd daemon on the specified address,
-the container stops immediately. For detailed information on working with this
-logging driver, see [the fluentd logging driver](fluentd.md)
+If container cannot connect to the Fluentd daemon on the specified address and
+`fluentd-async-connect` is not enabled, the container stops immediately.
+For detailed information on working with this logging driver,
+see [the fluentd logging driver](fluentd.md)
```
Comment (tagomoris) [inline docs/admin/logging/overview.md]: There's no details about these in fluentd.md, by removing 2 sections about `fail-on-startup-error` and `buffer-limit`.

**Comment by tagomoris** [comment]
I added some comments.
And I've just released `fluent-logger-golang` v1.1.0 with async connect and unix domain support options. It may be better version to be vendored.

**Comment by LK4D4** [comment]
@pcarrier feel free to update your PR.

**Comment by thaJeztah** [comment]
ping @pcarrier can you address the comments? Trying to move things forward here

**Comment by pcarrier** [comment]
On it.

File: docs/admin/logging/overview.md (hunk 5 above)
Comment (pcarrier) [inline docs/admin/logging/overview.md]: Yes sorry, I've kept documentation effort for status/3-docs-review.

**Comment by thaJeztah** [comment]
ping @cpuguy83 @LK4D4 PR's was updated, PTAL

@pcarrier will update the `fluentd.md` file with the new options during docs-review

(Thanks for updating @pcarrier!)

**Comment by icecrime** [comment]
test this please

**Comment by icecrime** [comment]
test this please

**Comment by icecrime** [comment]
Ping @SvenDowideit.

**Comment by thaJeztah** [comment]
retest this please

**Comment by SvenDowideit** [comment]
@pcarrier this PR needs a rebase - it still contains an old `docs/Dockerfile` that references a repo that is now private.

**Comment by pcarrier** [comment]
Done.

**Comment by thaJeztah** [comment]
Hm, networking issues? flaky test? @tiborvass?
https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22612/console
```
03:35:24 PASS: docker_cli_tag_test.go:171: DockerSuite.TestTagMatchesDigest	0.579s
03:35:24 PASS: docker_cli_tag_test.go:131: DockerSuite.TestTagOfficialNames	4.546s
03:35:46 
03:35:46 ----------------------------------------------------------------------
03:35:46 FAIL: docker_cli_tag_test.go:216: DockerSuite.TestTagTruncationAmbiguity
03:35:46 
03:35:46 docker_cli_tag_test.go:235:
03:35:46     id := inspectField(c, truncatedTag, "Id")
03:35:46 github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:1618:
03:35:46     ...open github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go: The system cannot find the path specified.
03:35:46 ... value *errors.errorString = &errors.errorString{s:"failed to inspect notbusybox:c8be56c709b8: \nAn error occurred trying to connect: Get https://45.55.33.7:2367/v1.23/containers/notbusybox:c8be56c709b8/json: dial tcp 45.55.33.7:2367: connectex: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.\n"} ("failed to inspect notbusybox:c8be56c709b8: \nAn error occurred trying to connect: Get https://45.55.33.7:2367/v1.23/containers/notbusybox:c8be56c709b8/json: dial tcp 45.55.33.7:2367: connectex: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.\n")
03:35:46 
03:35:48 
```

**Comment by calavera** [comment]
Why is this reverting 82f2f7041b4db3642e0fce1f2fd11234bee78f41 ?

**Comment by pcarrier** [comment]
> AFAICT this is because of 3cf82ff, which tries to solve the same problem in what I'd say isn't as good a solution. My solution would be to revert it.

Those conflicted. Which features from `3cf82ff` are missing here that I should bring back?

**Comment by calavera** [comment]
oh, sorry, I missed that comment.

LGTM

**Comment by cpuguy83** [comment]
We still need to support `fail-on-startup-error`, even if it's just an alias to the new option (no need for it in docs, though).
Can you update that and then we can move to docs review?

**Comment by pcarrier** [comment]
How would you like it to interact with `fluentd-async-connect`?

**Comment by cpuguy83** [comment]
I think if `fail-on-startup-error` is the inverse of `fluentd-async-connect`?

**Comment by thaJeztah** [comment]
ping @pcarrier what's the status here? Do you need more information?

**Comment by cpuguy83** [comment]
Looking at the `fail-on-startup-error` PR, maybe that didn't make 1.10?  https://github.com/docker/docker/commit/3cf82ff1ab14e1ddd2b629524e894ac359168388

So LGTM, moving to docs review.

File: docs/admin/logging/fluentd.md [hunk 6]
```diff
@@ -79,20 +78,6 @@ the log tag format.
 
 The `labels` and `env` options each take a comma-separated list of keys. If there is collision between `label` and `env` keys, the value of the `env` takes precedence. Both options add additional fields to the extra attributes of a logging message.
 
-### fail-on-startup-error
```
Comment (calavera) [inline docs/admin/logging/fluentd.md]: @pcarrier can we explain how `fluentd-async-connect` work here? I think that's enough to have the docs ready. I recommend you to rebase master at the same time :wink:

**Comment by pcarrier** [comment]
Sorry for the slow response, been busy with other tasks. On it.

File: docs/admin/logging/fluentd.md (hunk 6 above)
Comment (pcarrier) [inline docs/admin/logging/fluentd.md]: Done. For what it's worth I'm not sure I have another rebase in me :)

**Comment by pcarrier** [comment]
I don't understand.

```
% git log origin/master..pcarrier/19438-async-connections-to-fluentd
commit 7a29366b075f3b3cd7bc955a08afea2a6918c79c
Author: Pierre Carrier <pierre@meteor.com>
Date:   2016-03-20 16:30:46 +0000

    docs: fluentd-async-connect option
    
    Signed-off-by: Pierre Carrier <pierre@meteor.com>

commit d48eab75110f580022ff6af308241b4ce19957d1
Author: Pierre Carrier <pierre@meteor.com>
Date:   2016-01-24 05:51:03 -0800

    fluentd logger: support all options besides Unix sockets
    
    Mostly useful for docker/docker#19438.
    
    Signed-off-by: Pierre Carrier <pierre@meteor.com>

commit 9c2a0739fb9c5e3636425d669f0ab860ea1bfbae
Author: Pierre Carrier <pierre@meteor.com>
Date:   2016-01-19 00:07:13 -0800

    bump fluent-logger-golang version
    
    This version supports async connections.
    
    Signed-off-by: Pierre Carrier <pierre@meteor.com>

commit d89dae6e4becc16e80e3781ac68b9fbb855947b3
Author: Pierre Carrier <pierre@meteor.com>
Date:   2016-01-29 16:05:46 -0800

    Revert "Added flag to ignore fluentd connect error on container start"
    
    This reverts commit 3cf82ff1ab14e1ddd2b629524e894ac359168388.
    
    Signed-off-by: Pierre Carrier <pierre@meteor.com>
```

**Comment by icecrime** [comment]
Sorry @pcarrier, the commit do indeed look signed, I don't understand why the CI job is complaining (cc @thaJeztah). I guess we could ignore?

**Comment by thaJeztah** [comment]
docs LGTM, can you squash the last two commits?

**Comment by thaJeztah** [comment]
windows hitting flaky test https://github.com/docker/docker/issues/20922, restarting

**Comment by pcarrier** [comment]
@thaJeztah done :)

**Comment by thaJeztah** [comment]
cool, thanks, and looks like the "dco" check now also passed

LGTM

ping @vdemeester PTAL

**Comment by vdemeester** [comment]
LGTM 🐰

**Comment by thaJeztah** [comment]
Hm documentation fail looks to be a known issue in Hugo, because;

```
Found: 1119 files
Found: 0 errors
```

**Comment by thaJeztah** [comment]
retest this please

**Comment by thaJeztah** [comment]
gccgo is hitting https://github.com/docker/docker/issues/21365 :cry: restarting

**Pull request merged**